File: core/framework/test/suite/bridge/platform/features/animation.js
/*
 * Copyright (c) 2021, the hapjs-platform Project Contributors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import base from './base'

export const config = {
  name: 'system.animation',
  methods: [
    {
      name: 'enable',
      type: 0,
      mode: 0
    },
    {
      name: 'play',
      type: 0,
      mode: 0
    },
    {
      name: 'finish',
      type: 0,
      mode: 0
    },
    {
      name: 'pause',
      type: 0,
      mode: 0
    },
    {
      name: 'cancel',
      type: 0,
      mode: 0
    },
    {
      name: 'reverse',
      type: 0,
      mode: 0
    },
    {
      name: 'getFinished',
      type: 0,
      mode: 0
    },
    {
      name: 'setStartTime',
      type: 0,
      mode: 0
    },
    {
      name: 'getStartTime',
      type: 0,
      mode: 0
    },
    {
      name: 'getPlayState',
      type: 0,
      mode: 0
    }
  ]
}

const moduleOwn = Object.assign(
  {
    name: 'system.animation',
    _currentState: 'idle',
    _isFinished: 'false',
    _startTime: 0,

    enable(options = {}) {
      return this.mockSync(options._data, options._code)
    },

    play() {
      this._currentState = 'running'
      this._isFinished = 'false'
    },

    finish() {
      this._isFinished = 'true'
      this._currentState = 'finished'
    },

    pause() {
      this._currentState = 'paused'
      this._isFinished = 'false'
    },

    cancel() {
      this._isFinished = 'true'
      this._currentState = 'idle'
    },

    reverse() {
      this._currentState = 'running'
      this._isFinished = 'false'
    },

    getFinished() {
      return this.mockSync(this._isFinished)
    },

    getPlayState() {
      return this.mockSync(this._currentState)
    },

    setStartTime(params = {}) {
      this._startTime = params.startTime
    },

    getStartTime() {
      return this.mockSync(this._startTime)
    }
  },
  base
)

export default moduleOwn

File: core/runtime/android/runtime/src/main/java/org/hapjs/component/animation/Animation.java
/*
 * Copyright (c) 2021, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

package org.hapjs.component.animation;

import java.util.HashMap;
import java.util.Map;

import org.hapjs.component.Component;

public class Animation {
    private static final String TAG = "Animation";

    private int mRef;
    private CSSAnimatorSet mAnimatorSet;
    private Component mComponent;

    private OnCancelListener mCancelListener;
    private OnFinishListener mFinishListener;
    private Map<String, AnimationLifecycle> mLifecycleList = new HashMap<>();

    public Animation(CSSAnimatorSet animatorSet) {
        mAnimatorSet = animatorSet;
        mAnimatorSet.setAnimation(this);
    }

    public CSSAnimatorSet getAnimatorSet() {
        return mAnimatorSet;
    }

    public void setAnimatorSet(CSSAnimatorSet animatorSet) {
        mAnimatorSet = animatorSet;
    }

    public Component getComponent() {
        return mComponent;
    }

    public void setComponent(Component component) {
        mComponent = component;
    }

    public int getRef() {
        return mRef;
    }

    public void setRef(int ref) {
        mRef = ref;
    }

    public long getStartTime() {
        if (mAnimatorSet != null) {
            return mAnimatorSet.getStartTime();
        } else {
            return 0;
        }
    }

    public boolean isReady() {
        if (mAnimatorSet != null) {
            return mAnimatorSet.isReady();
        } else {
            return false;
        }
    }

    public boolean isFinished() {
        if (mAnimatorSet != null) {
            return mAnimatorSet.isFinished();
        } else {
            return true;
        }
    }

    public boolean isPending() {
        if (mAnimatorSet != null) {
            return mAnimatorSet.isPending();
        } else {
            return false;
        }
    }

    public String getPlayState() {
        if (mAnimatorSet != null) {
            return mAnimatorSet.getPlayState();
        } else {
            return "idle";
        }
    }

    public void play() {
        if (mAnimatorSet != null) {
            mAnimatorSet.start();
        }
    }

    public void pause() {
        if (mAnimatorSet != null) {
            mAnimatorSet.pause();
        }
    }

    public void resume() {
        if (mAnimatorSet != null) {
            mAnimatorSet.resume();
        }
    }

    public void finish() {
        if (mAnimatorSet != null) {
            mAnimatorSet.finish();
        }
    }

    public void cancel() {
        if (mAnimatorSet != null) {
            mAnimatorSet.cancel();
        }
    }

    public void reverse() {
        if (mAnimatorSet != null) {
            mAnimatorSet.reverse();
        }
    }

    public void onCancel() {
        if (mCancelListener != null) {
            mCancelListener.onCancel();
        }
    }

    public void onFinish() {
        if (mFinishListener != null) {
            mFinishListener.onFinish();
        }
    }

    public void onDestroy() {
        for (String key : mLifecycleList.keySet()) {
            AnimationLifecycle lifecycle = mLifecycleList.get(key);
            if (lifecycle != null) {
                lifecycle.onDestroy(key);
            }
        }
        if (mAnimatorSet != null) {
            // 先置空，避免互相调用 destroy，陷入循环
            CSSAnimatorSet animatorSet = mAnimatorSet;
            mAnimatorSet = null;
            animatorSet.destroy();
        }
    }

    public void setAnimationLifecycle(AnimationLifecycle lifecycle, String key) {
        mLifecycleList.put(key, lifecycle);
    }

    public void setCancelListener(OnCancelListener listener) {
        mCancelListener = listener;
    }

    public void setFinishListener(OnFinishListener listener) {
        mFinishListener = listener;
    }

    public interface OnCancelListener {
        void onCancel();
    }

    public interface OnFinishListener {
        void onFinish();
    }

    public interface AnimationLifecycle {
        void onDestroy(String key);
    }
}
